frontend/controller/serviceworkers/sw-primary.js

```diff
@@ -83,36 +235,82 @@ self.addEventListener('message', function (event) {
+      clientList.sort((a, b) => {
+        if (a.focused !== b.focused) {
+          return a.focused ? -1 : 1
+        }
+        if (a.visibilityState !== b.visibilityState) {
+        // order is visible, prerender, hidden
+          if (a.visibilityState === 'visible') return -1
+          if (b.visibilityState === 'visible') return 1
+          if (a.visibilityState === 'hidden') return 1
+          if (b.visibilityState === 'hidden') return -1
+        }
 
-  const data = event.data.json()
-  console.debug('[sw] push received: ', data)
+        return 0
+      })
+      if (!clientList.length) {
+        return self.clients.openWindow(`${sbp('controller/router').options.base}${event.notification.data.path ?? '/'}`)
+      }
+      const client = clientList[0]
+      if (event.notification.data?.path) {
+        client.postMessage({
+          type: 'navigate',
+          path: event.notification.data.path
+        })
+      }
+      if (!client.focused) return client.focus()
+    }))
+})
 
-  self.registration.showNotification(
-    data.title,
-    {
-      body: data.body || '',
-      icon: '/assets/images/pwa-icons/group-income-icon-transparent.svg'
+sbp('okTurtles.events/on', KV_EVENT, ({ contractID, key, data }) => {
+  const rootState = sbp('chelonia/rootState')
+  const ourIdentityContractID = rootState.loggedIn?.identityContractID
+  if (contractID !== ourIdentityContractID) return
+  // the following keys mirror the corresponding keys in Vuex modules in the
+  // app
+  switch (key) {
+    case KV_KEYS.LAST_LOGGED_IN: {
+      if (!rootState.lastLoggedIn) rootState.lastLoggedIn = Object.create(null)
+      rootState.lastLoggedIn[contractID] = data
+      break
+    }
+    case KV_KEYS.UNREAD_MESSAGES: {
+      if (!rootState.chatroom) rootState.chatroom = Object.create(null)
+      rootState.chatroom.unreadMessages = data
+      break
+    }
+    case KV_KEYS.PREFERENCES: {
+      rootState.preferences = data
+      break
+    }
+    case KV_KEYS.NOTIFICATIONS: {
+      if (!rootState.notifications) rootState.notifications = Object.create(null)
+      rootState.notifications.status = data
+      break
     }
-  )
+    default:
+      return
+  }
+  sbp('okTurtles.events/emit', CHELONIA_STATE_MODIFIED)
 })
 
-self.addEventListener('pushsubscriptionchange', async function (event) {
-  // NOTE: Currently there is no specific way to validate if a push-subscription is valid. So it has to be handled in the front-end.
-  // (reference:https://pushpad.xyz/blog/web-push-how-to-check-if-a-push-endpoint-is-still-valid)
-  const subscription = await self.registration.pushManger.subscribe(event.oldSubscription.options)
+sbp('okTurtles.events/on', NEW_CHATROOM_UNREAD_POSITION, ({ chatRoomID, messageHash }) => {
+  const rootState = sbp('chelonia/rootState')
+  if (messageHash) {
+    // `rootState.chatroom` mirrors the 'chatroom' module in the app
+    if (!rootState.chatroom) rootState.chatroom = Object.create(null)
```

Comment:
I'm noticing quite a lot of repetitive code here for touching stuff on the `rootState`.

There's this check, then the check below on line 311, and way above on line 283.

Is there a way to better organize and group all of this state initialization somewhere? Perhaps near the top of the file?

Is `sbp('chelonia/rootState')` available when this file loads? If not, when is it first available and can the initialization be done there or near there?

Then we would need `if (!rootState.chatroom)`, `if (!rootState.lastLoggedIn)`, `if (!rootState.notifications)`, `if (!rootState.chatroom?.chatRoomScrollPosition)` etc. all over the place.

Reply:
> Is sbp('chelonia/rootState') available when this file loads?

It should be. The issue is that calling `chelonia/reset` can and will delete these entries from the state.

Reply:
In that case, can you put all this initialization code in a function that gets called once at the start/top of this file and then again inside of an event listener for an event that `chelonia/reset` emits at the end of its doings?